src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/PublicContract.cs

```diff
@@ -0,0 +1,162 @@
+                ThrowArgumentItemNullOrDuplicateException(sequence, argumentName);
+            }
+        }
+
+        /// <summary>
+        /// Use to validate public API input for properties that are exposed as <see cref="IReadOnlyList{T}"/>.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static IReadOnlyList<T> ToBoxedImmutableArrayWithNonNullItems<T>(this IEnumerable<T>? sequence, string argumentName) where T : class
+        {
+            var list = sequence.ToBoxedImmutableArray();
+
+            if (list.Contains((T)null!))
+            {
+                ThrowArgumentItemNullException(list, argumentName);
+            }
+
+            return list;
+        }
+
+        /// <summary>
+        /// Use to validate public API input for properties that are exposed as <see cref="IReadOnlyList{T}"/> and 
+        /// whose items should be unique.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static IReadOnlyList<T> ToBoxedImmutableArrayWithDistinctNonNullItems<T>(this IEnumerable<T>? sequence, string argumentName) where T : class
+        {
+            var list = sequence.ToBoxedImmutableArray();
+
+            if (list.IndexOfNullOrDuplicateItem() >= 0)
+            {
+                ThrowArgumentItemNullOrDuplicateException(list, argumentName);
+            }
+
+            return list;
+        }
+
+        private static int IndexOfNullOrDuplicateItem<T>(this IEnumerable<T> sequence) where T : class
+            => (sequence is IReadOnlyList<T> list) ? IndexOfNullOrDuplicateItem(list) : EnumeratingIndexOfNullOrDuplicateItem(sequence);
+
+        private static int EnumeratingIndexOfNullOrDuplicateItem<T>(IEnumerable<T> sequence) where T : class
+        {
+            using var _ = PooledHashSet<T>.GetInstance(out var set);
+
+            int i = 0;
+            foreach (var item in sequence)
+            {
+                if (item is null || !set.Add(item))
+                {
+                    return i;
+                }
+
+                i++;
+            }
+
+            return -1;
+        }
+
+        private static int IndexOfNullOrDuplicateItem<T>(this IReadOnlyList<T> list) where T : class
+        {
+            int length = list.Count;
+
+            if (length == 0)
+            {
+                return -1;
+            }
+
+            if (length == 1)
+            {
+                return (list[0] is null) ? 0 : -1;
+            }
+
+            using var _ = PooledHashSet<T>.GetInstance(out var set);
+
+            for (int i = 0; i < length; i++)
+            {
+                var item = list[i];
+                if (item is null || !set.Add(item))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private static string MakeIndexedArgumentName(string argumentName, int index)
+            => $"{argumentName}[{index}]";
```

Comment:
Is this locale sensitive in some way?

Reply:
I don't think it matters.

Reply:
This is just a helpful hint to the developer.

Reply:
It's all a helpful hint to the developer until somebody writes exception catch filters based on the exception message. :smile: